File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/infrastructure/geometry/Vector.kt
package com.crescentflare.slicinggame.infrastructure.geometry

import android.graphics.PointF
import android.graphics.RectF
import kotlin.math.*

/**
 * Geometry: defines a vector with a start/end position and direction
 */
class Vector {

    // --
    // Members
    // --

    var x: Float
        get() = end.x - start.x
        set(x) {
            end.x = start.x + x
        }

    var y: Float
        get() = end.y - start.y
        set(y) {
            end.y = start.y + y
        }

    var start: PointF
    var end: PointF


    // --
    // Initialization
    // --

    constructor(x: Float, y: Float): this(PointF(), PointF(x, y))

    constructor(start: PointF, end: PointF) {
        this.start = start
        this.end = end
    }

    constructor(directionAngle: Float, pivot: PointF? = null) {
        this.start = pivot ?: PointF()
        this.end = PointF(sin(start.x + directionAngle * PI.toFloat() * 2 / 360), start.y - cos(directionAngle * PI.toFloat() * 2 / 360))
    }


    // --
    // Operators
    // --

    operator fun times(multiplier: Float): Vector {
        return Vector(PointF(start.x, start.y), PointF(start.x + x * multiplier, start.y + y * multiplier))
    }

    operator fun timesAssign(multiplier: Float) {
        x *= multiplier
        y *= multiplier
    }


    // --
    // Return modified result
    // --

    fun unit(): Vector {
        val magnitude = distance()
        return Vector(x / magnitude, y / magnitude)
    }

    fun translated(translateX: Float, translateY: Float): Vector {
        return Vector(PointF(start.x + translateX, start.y + translateY), PointF(end.x + translateX, end.y + translateY))
    }

    fun scaled(scaleX: Float, scaleY: Float): Vector {
        return Vector(PointF(start.x * scaleX, start.y * scaleY), PointF(end.x * scaleX, end.y * scaleY))
    }

    fun reversed(): Vector {
        return Vector(end, start)
    }

    fun perpendicular(): Vector {
        return Vector(PointF(start.x, start.y), PointF(start.x + y, start.y - x))
    }

    fun stretchedToEdges(topLeft: PointF, bottomRight: PointF): Vector {
        val newStart = PointF(start.x, start.y)
        val newEnd = PointF(end.x, end.y)
        if (abs(start.x - end.x) > abs(start.y - end.y)) {
            val slope = (start.y - end.y) / (start.x - end.x)
            if (start.x < end.x) {
                newStart.x = topLeft.x
                newStart.y += (topLeft.x - start.x) * slope
                newEnd.x = bottomRight.x
                newEnd.y += (bottomRight.x - end.x) * slope
            } else {
                newStart.x = bottomRight.x
                newStart.y += (bottomRight.x - start.x) * slope
                newEnd.x = topLeft.x
                newEnd.y += (topLeft.x - end.x) * slope
            }
        } else {
            val slope = (start.x - end.x) / (start.y - end.y)
            if (start.y < end.y) {
                newStart.x += (topLeft.y - start.y) * slope
                newStart.y = topLeft.y
                newEnd.x += (bottomRight.y - end.y) * slope
                newEnd.y = bottomRight.y
            } else {
                newStart.x += (bottomRight.y - start.y) * slope
                newStart.y = bottomRight.y
                newEnd.x += (topLeft.y - end.y) * slope
                newEnd.y = topLeft.y
            }
        }
        return Vector(newStart, newEnd)
    }


    // --
    // Checks
    // --

    fun isValid(): Boolean {
        return start.x != end.x || start.y != end.y
    }

    fun distance(): Float {
        return sqrt(x * x + y * y)
    }

    fun directionOfPoint(point: PointF): Float {
        return (point.y - end.y) * (end.x - start.x) - (end.y - start.y) * (point.x - end.x)
    }

    fun intersect(otherVector: Vector): PointF? {
        // Intersection direction formula
        val d = x * otherVector.y - y * otherVector.x
        if (d == 0f) {
            return null
        }

        // Distance formulas
        val u = ((otherVector.start.x - start.x) * (otherVector.end.y - otherVector.start.y) - (otherVector.start.y - start.y) * (otherVector.end.x - otherVector.start.x)) / d
        val v = ((otherVector.start.x - start.x) * (end.y - start.y) - (otherVector.start.y - start.y) * (end.x - start.x)) / d
        if (u < 0 || u > 1) {
            return null
        }
        if (v < 0 || v > 1) {
            return null
        }

        // Return the intersection result
        return PointF(start.x + u * (end.x - start.x), start.y + u * (end.y - start.y))
    }

    fun intersect(rect: RectF): PointF? {
        // Return early if the starting point is already inside the rectangle
        if (rect.contains(start.x, start.y)) {
            return PointF(start.x, start.y)
        }

        // Calculate intersection distances for each axis separately
        val entryDistanceX = if (x > 0) rect.left - start.x else rect.right - start.x
        val entryDistanceY = if (y > 0) rect.top - start.y else rect.bottom - start.y
        val exitDistanceX = if (x > 0) rect.right - start.x else rect.left - start.x
        val exitDistanceY = if (y > 0) rect.bottom - start.y else rect.top - start.y

        // Calculate intersection time relative to the vector length (ranging from 0 to 1)
        val entryTimeX = if (x == 0f) Float.POSITIVE_INFINITY else entryDistanceX / x
        val entryTimeY = if (y == 0f) Float.POSITIVE_INFINITY else entryDistanceY / y
        val exitTimeX = if (x == 0f) Float.POSITIVE_INFINITY else exitDistanceX / x
        val exitTimeY = if (y == 0f) Float.POSITIVE_INFINITY else exitDistanceY / y

        // Check for intersection and return result
        val entryTime = max(entryTimeX, entryTimeY)
        val exitTime = min(exitTimeX, exitTimeY)
        if (entryTime < exitTime && entryTime >= 0 && entryTime <= 1) {
            return PointF(start.x + x * entryTime, start.y + y * entryTime)
        }
        return null
    }

    fun intersect(polygon: Polygon): PointF? {
        // Return early if the starting point is already inside the polygon
        if (polygon.contains(start)) {
            return PointF(start.x, start.y)
        }

        // Check if intersecting with the polygon lines
        var closestIntersection: PointF? = null
        for (vector in polygon.asVectorList()) {
            if (vector.directionOfPoint(start) < 0 && vector.directionOfPoint(end) > 0) {
                vector.intersect(this)?.let {
                    val previousIntersection = closestIntersection
                    if (previousIntersection != null) {
                        if (abs(it.x - vector.start.x) < abs(previousIntersection.x - vector.start.x) && abs(it.y - vector.start.y) < abs(previousIntersection.y - vector.start.y)) {
                            closestIntersection = it
                        }
                    } else {
                        closestIntersection = it
                    }
                }
            }
        }
        return closestIntersection
    }

    fun edgeIntersections(polygon: Polygon): List<PointF> {
        val points = mutableListOf<PointF>()
        for (vector in polygon.asVectorList()) {
            intersect(vector)?.let {
                points.add(it)
            }
        }
        return points
    }

}

File: slicing-ios/SlicingGame/Infrastructure/Geometry/Vector.swift
//
//  Polygon.swift
//  Geometry: defines a vector with a start/end position and direction
//

import UIKit

class Vector {

    // --
    // MARK: Members
    // --
    
    var x: CGFloat {
        set {
            end.x = start.x + newValue
        }
        get { return end.x - start.x }
    }

    var y: CGFloat {
        set {
            end.y = start.y + newValue
        }
        get { return end.y - start.y }
    }

    var start = CGPoint()
    var end = CGPoint()


    // --
    // MARK: Initialization
    // --
    
    convenience init(x: CGFloat, y: CGFloat) {
        self.init(start: CGPoint.zero, end: CGPoint(x: x, y: y))
    }
    
    init(start: CGPoint, end: CGPoint) {
        self.start = start
        self.end = end
    }

    init(directionAngle: CGFloat, pivot: CGPoint? = nil) {
        self.start = pivot ?? CGPoint.zero
        self.end = CGPoint(x: start.x + sin(directionAngle * CGFloat.pi * 2 / 360), y: start.y - cos(directionAngle * CGFloat.pi * 2 / 360))
    }

    
    // --
    // MARK: Operators
    // --

    static func *(vector: Vector, multiplier: CGFloat) -> Vector {
        return Vector(start: vector.start, end: CGPoint(x: vector.start.x + vector.x * multiplier, y: vector.start.y + vector.y * multiplier))
    }
    
    static func *=(vector: Vector, multiplier: CGFloat) {
        vector.x *= multiplier
        vector.y *= multiplier
    }
    

    // --
    // MARK: Return modified result
    // --
    
    func unit() -> Vector {
        let magnitude = distance()
        return Vector(x: x / magnitude, y: y / magnitude)
    }
    
    func translated(translateX: CGFloat, translateY: CGFloat) -> Vector {
        return Vector(start: CGPoint(x: start.x + translateX, y: start.y + translateY), end: CGPoint(x: end.x + translateX, y: end.y + translateY))
    }
    
    func scaled(scaleX: CGFloat, scaleY: CGFloat) -> Vector {
        return Vector(start: CGPoint(x: start.x * scaleX, y: start.y * scaleY), end: CGPoint(x: end.x * scaleX, y: end.y * scaleY))
    }

    func reversed() -> Vector {
        return Vector(start: end, end: start)
    }
    
    func perpendicular() -> Vector {
        return Vector(start: start, end: CGPoint(x: start.x + y, y: start.y - x))
    }

    func stretchedToEdges(topLeft: CGPoint, bottomRight: CGPoint) -> Vector {
        var newStart = start
        var newEnd = end
        if (abs(start.x - end.x) > abs(start.y - end.y)) {
            let slope = (start.y - end.y) / (start.x - end.x)
            if (start.x < end.x) {
                newStart.x = topLeft.x
                newStart.y += (topLeft.x - start.x) * slope
                newEnd.x = bottomRight.x
                newEnd.y += (bottomRight.x - end.x) * slope
            } else {
                newStart.x = bottomRight.x
                newStart.y += (bottomRight.x - start.x) * slope
                newEnd.x = topLeft.x
                newEnd.y += (topLeft.x - end.x) * slope
            }
        } else {
            let slope = (start.x - end.x) / (start.y - end.y)
            if (start.y < end.y) {
                newStart.x += (topLeft.y - start.y) * slope
                newStart.y = topLeft.y
                newEnd.x += (bottomRight.y - end.y) * slope
                newEnd.y = bottomRight.y
            } else {
                newStart.x += (bottomRight.y - start.y) * slope
                newStart.y = bottomRight.y
                newEnd.x += (topLeft.y - end.y) * slope
                newEnd.y = topLeft.y
            }
        }
        return Vector(start: newStart, end: newEnd)
    }

    
    // --
    // MARK: Checks
    // --
    
    func isValid() -> Bool {
        return start.x != end.x || start.y != end.y
    }
    
    func distance() -> CGFloat {
        return sqrt(x * x + y * y)
    }

    func directionOf(point: CGPoint) -> CGFloat {
        return (point.y - end.y) * (end.x - start.x) - (end.y - start.y) * (point.x - end.x)
    }
    
    func intersect(withVector: Vector) -> CGPoint? {
        // Intersection direction formula
        let d = x * withVector.y - y * withVector.x
        if d == 0 {
            return nil
        }
        
        // Distance formulas
        let u = ((withVector.start.x - start.x) * (withVector.end.y - withVector.start.y) - (withVector.start.y - start.y) * (withVector.end.x - withVector.start.x)) / d
        let v = ((withVector.start.x - start.x) * (end.y - start.y) - (withVector.start.y - start.y) * (end.x - start.x)) / d
        if u < 0 || u > 1 {
            return nil
        }
        if v < 0 || v > 1 {
            return nil
        }
        
        // Return the intersection result
        return CGPoint(x: start.x + u * (end.x - start.x), y: start.y + u * (end.y - start.y))
    }
    
    func intersect(withRect: CGRect) -> CGPoint? {
        // Return early if the starting point is already inside the rectangle
        if withRect.contains(start) {
            return CGPoint(x: start.x, y: start.y)
        }
        
        // Calculate intersection distances for each axis separately
        let entryDistanceX = x > 0 ? withRect.minX - start.x : withRect.maxX - start.x
        let entryDistanceY = y > 0 ? withRect.minY - start.y : withRect.maxY - start.y
        let exitDistanceX = x > 0 ? withRect.maxX - start.x : withRect.minX - start.x
        let exitDistanceY = y > 0 ? withRect.maxY - start.y : withRect.minY - start.y
        
        // Calculate intersection time relative to the vector length (ranging from 0 to 1)
        let entryTimeX = x == 0 ? CGFloat.infinity : entryDistanceX / x
        let entryTimeY = y == 0 ? CGFloat.infinity : entryDistanceY / y
        let exitTimeX = x == 0 ? CGFloat.infinity : exitDistanceX / x
        let exitTimeY = y == 0 ? CGFloat.infinity : exitDistanceY / y
        
        // Check for intersection and return result
        let entryTime = max(entryTimeX, entryTimeY)
        let exitTime = min(exitTimeX, exitTimeY)
        if entryTime < exitTime && entryTime >= 0 && entryTime <= 1 {
            return CGPoint(x: start.x + x * entryTime, y: start.y + y * entryTime)
        }
        return nil
    }
    
    func intersect(withPolygon: Polygon) -> CGPoint? {
        // Return early if the starting point is already inside the polygon
        if withPolygon.contains(start) {
            return CGPoint(x: start.x, y: start.y)
        }
        
        // Check if intersecting with the polygon lines
        var closestIntersection: CGPoint?
        for vector in withPolygon.asVectorArray() {
            if vector.directionOf(point: start) < 0 && vector.directionOf(point: end) > 0 {
                if let intersection = vector.intersect(withVector: self) {
                    if let previousIntersection = closestIntersection {
                        if abs(intersection.x - vector.start.x) < abs(previousIntersection.x - vector.start.x) && abs(intersection.y - vector.start.y) < abs(previousIntersection.y - vector.start.y) {
                            closestIntersection = intersection
                        }
                    } else {
                        closestIntersection = intersection
                    }
                }
            }
        }
        return closestIntersection
    }
    
    func edgeIntersections(withPolygon: Polygon) -> [CGPoint] {
        var points = [CGPoint]()
        for vector in withPolygon.asVectorArray() {
            if let point = intersect(withVector: vector) {
                points.append(point)
            }
        }
        return points
    }

}
